### PR TITLE
Frontend/feature/expired transfers

### LINKED ISSUE #606: Allow users to withdraw expired requests
# Rationale
In the unhappy case that a user creates a request that is not filled (for whatever reason) within the timeout period, we need to have a way in the UI to allow the user to withdraw the tokens that are locked with his request. The functionality is already implemented in the smart contracts, but not in the frontend.

# Implementation
I believe it would be ideal to offer this functionality as part of the "transaction history" overview, so that the user can see if a request wasn't successful and has the options to press a button "withdraw funds" to get back his funds.

I need input from @olympe-silve regarding this and it would be great if it could be included in the design of the transaction history. I am also open to other suggestions, but this currently the best place for this functionality in my opinion.

### PR DESCRIPTION
**Pull request by weilbith**
Closes #606
Based on #708 (though not required, just avoiding conflicts).

See the commit messages for important details.
The first few commits are general improvements in preparation before the actual feature implementation.

### CODE REVIEW

**Comment by weilbith** [comment]
As #708 got merged, I can go ahead and rebase on top of `main` to remove the redundant commits.

**Comment by manuelwedler** [comment]
Please do so :+1:

**Comment by weilbith** [comment]
Note to myself: double check nothing gets lost while we merge more visual related pull requests in the meantime. The component renaming was pretty nasty in this sense.

File: frontend/src/actions/transfers/transfer.ts
Comment (manuelwedler) [inline frontend/src/actions/transfers/transfer.ts:138]: Does the page really reload? I think in the other places the reload is manually triggered after the `switchChain` call. We could also change the provider implementation to do it there.

**Review by manuelwedler — COMMENTED** [review]
Thanks for the good work! Looks like a high quality PR again :)

File: frontend/src/services/transactions/fill-manager.ts
Comment (weilbith) [reply frontend/src/services/transactions/fill-manager.ts:27]: Okay, I can do that. This will be refactored soon anyway, but I like such recommendations and thereby follow it. Always good to keep the code base clean, even if you plan to remove stuff soonish.

File: frontend/src/services/transactions/request-manager.ts
Comment (weilbith) [reply frontend/src/services/transactions/request-manager.ts:20]: Okay, I'll do. I forgot that this is our provider implementation and not `ethers`. Sometimes confusing. 🙈

File: frontend/src/actions/transfers/transfer.ts
Comment (weilbith) [reply frontend/src/actions/transfers/transfer.ts:138]: We observe the chain identifier and reload it [here](https://github.com/beamer-bridge/beamer/blob/2f251fae7cebfc328b6ceabaeecaf0f7fa3eacac/frontend/src/components/RequestDialog.vue#L126).
I kinda don't like this whole page reloading on chain change etc. I'm not 100% sure if it is actually necessary at all. So I guess I tried to minimize this effect application.
Anyway, I think you have a good point and I can ensure that this happens at the `EthereumProvider`. There it is really close to the source of event + change and no Vue component has to take care of that.

File: frontend/tests/unit/components/Transfer.spec.ts
Comment (manuelwedler) [reply frontend/tests/unit/components/Transfer.spec.ts:207]: Ok good point about the indices. I like that syntax actually better, too :)

File: frontend/src/services/transactions/request-manager.ts
Comment (weilbith) [reply frontend/src/services/transactions/request-manager.ts:20]: While attempting to do it: No, we use it at a couple of other places like the token too.

File: frontend/tests/unit/components/Transfer.spec.ts
Comment (weilbith) [reply frontend/tests/unit/components/Transfer.spec.ts]: ✔️

File: frontend/tests/unit/actions/transfers/transfer.spec.ts
Comment (weilbith) [reply frontend/tests/unit/actions/transfers/transfer.spec.ts]: ✔️

File: frontend/src/actions/transfers/transfer.ts
Comment (weilbith) [reply frontend/src/actions/transfers/transfer.ts:138]: > There was some error that made us introduce the location reload thing. I could find the related PR: #388

I also read in the [MetaMask](https://docs.metamask.io/guide/ethereum-provider.html#chainchanged) docs that it is strongly recommended. I still would love to know the exact reasoning. 
Thank you very much or the link!

File: frontend/src/services/transactions/fill-manager.ts
Comment (weilbith) [reply frontend/src/services/transactions/fill-manager.ts]: It was less effective than I though due to the mixmax with signers and sometimes functions need the contract but also the provider. Anyway, I did it.

Comment (weilbith) [reply frontend/src/services/transactions/fill-manager.ts]: ✔️

File: frontend/src/actions/transfers/transfer.ts
Comment (weilbith) [reply frontend/src/actions/transfers/transfer.ts:138]: I think I'm mixing up two things here. Reload when we change the source chain and reloading when the wallet changed the network.
The first part I want to change in future. The second is already solved properly.
Thereby I'm closing. 👍🏾

**Comment by weilbith** [comment]
Alright, PR is clean an mergable now.

@manuelwedler please check if you're satisfied with the changes and approve if so (use the fixup commits in the PR history as support).

I'll do a final practical test to see if it still works. That needs 10min to have the transfer expire.

**Comment by weilbith** [comment]
Works. PR is fine from my site. ✔️

**Review by manuelwedler — APPROVED** [review]
Looks great!

**Comment by weilbith** [comment]
Currently the state `withdrawn` is exclusively set by us internally. Having this, we need to consider if it is possible to extend it with the knowledge of the chain.

**Comment by weilbith** [comment]
Bug is fixed now with some additional improvement related to it. Commit message includes the details.

**Comment by weilbith** [comment]
Alright, I did a long duration test to verify the bug fix. So I disabled the detection of fulfillment to strain the expiration detection, added some logging and then tested it with and without the fix. For both setups once with an actual expiring transfer and one that got fulfilled (though we do not detect that on purpose).
Without the fix, the transfers gets incorrectly detected as expired when the claim period ends and the agent claims the deposit. Doing a withdraw now would fail. Though in reality this would have went fine as the fulfillment event was happening an hour earlier.
The fixed version was able to prove that it does recognize the deposit got claimed by an agent and did not expire. 👍🏾

**Comment by weilbith** [comment]
I got already the review by Manuel and all feedback got discussed and addressed.
As this PR is currently blocking two others, we are under pressure for the next testnet release to being able introducing the breaking change for the new contracts, I decide to merge this PR now. The additional changes are not very drastic. Manuel will work only half a day next week and also not on Monday. So as for many of the prior PRs, the current personnel situation isn't easy. I got a review, it covers 90% of the PRs final state. That must be good enough for now.

**Pull request merged**
